lizard_auth_server/views_api_v2.py

```diff
@@ -840,3 +844,122 @@ def form_valid(self, form):
+
+        A 409 status if there are multiple users with given username/email
+        (case insensitive). A warning will be logged in this case.
+        """
+        # The JWT message is validated; now check the message's contents.
+        username = form.cleaned_data.get("username")
+        if not username:
+            return HttpResponseBadRequest("username is missing from the JWT message")
+
+        portal = Portal.objects.get(sso_key=form.cleaned_data["iss"])
+        if not portal.allow_migrate_user:
+            raise PermissionDenied("this portal is not allowed to migrate users")
+
+        # Do the authentication without the django backends, because we do not
+        # want to migrate LDAP user and we certainly do not want to do a call
+        # to Cognito, else we end up in an infinite loop.
+        try:
+            user = User.objects.get(
+                Q(username__iexact=username) | Q(email__iexact=username),
+                is_active=True,
+                user_profile__migrated_at=None,
+            )
+        except User.DoesNotExist:
+            return HttpResponseNotFound("No user found")
+        except User.MultipleObjectsReturned:
+            logger.warning("Multiple users found with username/email %s", username)
+            return HttpResponse("Multiple users found", status=409)
+
+        # Verify the password, if supplied
+        password = form.cleaned_data.get("password")
+        if password is None:
+            # Forgot password flow
+            password_valid = False  # ignored
+            migrate = True
+        else:
+            # Authentication flow
+            password_valid = user.check_password(password)
+            migrate = password_valid
+
+        # Record this call
+        if migrate:
+            UserProfile.objects.filter(user=user).update(migrated_at=timezone.now())
+
+        data = {
+            "user": construct_user_data(user=user),
+            "password_valid": password_valid,
+        }
+        return JsonResponse(data)
+
+
+class CognitoUserExistsView(CheckCredentialsView):
+    """View to check user existence for AWS Cognito
+
+    This view accepts "username" and (optionally) "email" in the request.
+
+    A user "exists" if the username already is present as username or email in
+    the local SSO database.
+    If email is present in the request, it is also checked if the email is
+    present (as username and email)
+
+    All checks are case-insensitive and inactive / migrated users are ignored.
+    """
+
+    def form_valid(self, form):
+        """Check for user existence.
+
+        Returns:
+          A dict with keys
+          - ``exists`` bool
+        """
+        # The JWT message is validated; now check the message's contents.
+        username = form.cleaned_data.get("username")
+        email = form.cleaned_data.get("email")
+        if not username:
+            return HttpResponseBadRequest("username is missing from the JWT message")
```

Comment:
Hier is username verplicht. In de andere view wordt er naar username en/of email gekeken.
Is dat hoe (onze functie bij) cognito het verwacht?

Reply:
Ja, de gewone authenticatie heeft 1 username (die zowel username als email kan betekenen)

De signup heeft username en email, die beiden niet mogen clashen met bestaande username/email.